# Final Project: Breast Cancer Prediction Model using SVM

### Candidate Name: Mayur Kumar Sharma
### Batch: DW-32

In [67]:
# Using the given dataset, I've created a model which is useful for breast cancer prediction.

### Importing the required libraries & packages

In [59]:
import pandas as pd
import numpy as np
import sklearn.datasets
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn import svm
from sklearn.metrics import classification_report

In [17]:
# Reading the dataset & creating a dataframe

df = pd.read_csv('cancer.csv')
df

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         M        17.99         10.38          122.80     1001.0   
1      842517         M        20.57         17.77          132.90     1326.0   
2    84300903         M        19.69         21.25          130.00     1203.0   
3    84348301         M        11.42         20.38           77.58      386.1   
4    84358402         M        20.29         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
564    926424         M        21.56         22.39          142.00     1479.0   
565    926682         M        20.13         28.25          131.20     1261.0   
566    926954         M        16.60         28.08          108.30      858.1   
567    927241         M        20.60         29.33          140.10     1265.0   
568     92751         B         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0    ...          17.33           184.60      2019.0           0.16220   
1    ...          23.41           158.80      1956.0           0.12380   
2    ...          25.53           152.50      1709.0           0.14440   
3    ...          26.50            98.87       567.7           0.20980   
4    ...          16.67           152.20      1575.0           0.13740   
..   ...            ...              ...         ...               ...   
564  ...          26.40           166.10      2027.0           0.14100   
565  ...          38.25           155.00      1731.0           0.11660   
566  ...          34.12           126.70      1124.0           0.11390   
567  ...          39.42           184.60      1821.0           0.16500   
568  ...          30.37            59.16       268.6           0.08996   

     compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0              0.66560           0.7119                0.2654          0.4601   
1              0.18660           0.2416                0.1860          0.2750   
2              0.42450           0.4504                0.2430          0.3613   
3              0.86630           0.6869                0.2575          0.6638   
4              0.20500           0.4000                0.1625          0.2364   
..                 ...              ...                   ...             ...   
564            0.21130           0.4107                0.2216          0.2060   
565            0.19220           0.3215                0.1628          0.2572   
566            0.30940           0.3403                0.1418          0.2218   
567            0.86810           0.9387                0.2650          0.4087   
568            0.06444           0.0000                0.0000          0.2871   

     fractal_dimension_worst  Unnamed: 32  
0                    0.11890          NaN  
1                    0.08902          NaN  
2                    0.08758          NaN  
3                    0.17300          NaN  
4      

In [18]:
df.shape

(569, 33)

In [19]:
df.describe()

id  radius_mean  texture_mean  perimeter_mean    area_mean  \
count  5.690000e+02   569.000000    569.000000      569.000000   569.000000   
mean   3.037183e+07    14.127292     19.289649       91.969033   654.889104   
std    1.250206e+08     3.524049      4.301036       24.298981   351.914129   
min    8.670000e+03     6.981000      9.710000       43.790000   143.500000   
25%    8.692180e+05    11.700000     16.170000       75.170000   420.300000   
50%    9.060240e+05    13.370000     18.840000       86.240000   551.100000   
75%    8.813129e+06    15.780000     21.800000      104.100000   782.700000   
max    9.113205e+08    28.110000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       symmetry_mean  ...  texture_worst  perimeter_worst   area_worst  \
count     569.000000  ...     569.000000       569.000000   569.000000   
mean        0.181162  ...      25.677223       107.261213   880.583128   
std         0.027414  ...       6.146258        33.602542   569.356993   
min         0.106000  ...      12.020000        50.410000   185.200000   
25%         0.161900  ...      21.080000        84.110000   515.300000   
50%         0.179200  ...      25.410000        97.660000   686.500000   
75%         0.195700  ...      29.720000       125.400000  1084.000000   
max         0.304000  ...      49.540000       251.200000  4254.000000   

       smoothness_worst  compactness_worst  concavity_worst  \
count        569.000000         569.000000       569.000000   
mean           0.132369           0.254265         0.272188   
std            0.022832           0.157336         0.208624   
min            0.071170           0.027290         0.000000   
25%            0.116600           0.147200         0.114500   
50%            0.131300           0.211900         0.226700   
75%            0.146000           0.339100         0.382900   
max            0.222600           1.058000         1.252000   

       concave points_worst  symmetry_worst  fractal_dimension_worst  \
count            569.000000      569.000000               569.000000   
mean               0.114606        0.290076                 0.083946   
std                0.065732        0.061867                 0.018061   
min                0.000000        0.156500                 0.055040   
25%                0.064930        0.250400                 0.071460   
50%                0.099930        0.282200                 0.080040   
75%                0.161400        0.317900                 0.092080   
max                0.291000        0.663800                 0.207500   

       Unnamed: 32  
count          0.0  
mean           NaN  
std            NaN  
min            NaN  
25%            NaN  
50%            NaN  
75%            NaN  
max            NaN  

[8 rows x 32 columns]

By using describe() method we get useful information about the dataset as shown above.

In [20]:
df.dtypes

id                           int64
diagnosis                   object
radius_mean                float64
texture_mean               float64
perimeter_mean             float64
area_mean                  float64
smoothness_mean            float64
compactness_mean           float64
concavity_mean             float64
concave points_mean        float64
symmetry_mean              float64
fractal_dimension_mean     float64
radius_se                  float64
texture_se                 float64
perimeter_se               float64
area_se                    float64
smoothness_se              float64
compactness_se             float64
concavity_se               float64
concave points_se          float64
symmetry_se                float64
fractal_dimension_se       float64
radius_worst               float64
texture_worst              float64
perimeter_worst            float64
area_worst                 float64
smoothness_worst           float64
compactness_worst          float64
concavity_worst     

In [21]:
df['diagnosis']

0      M
1      M
2      M
3      M
4      M
      ..
564    M
565    M
566    M
567    M
568    B
Name: diagnosis, Length: 569, dtype: object

By checking the data-types of columns, we can find the type of data present in the respective feature.
Here 'diagnosis' has object data-type, on which we can not perform mathematical operations. 
So we'll encode it using Label-Encoder technique. 
(It has two classes 'M' & 'B')
Then we'll drop this feature & add new feature 'result', obtained from label-encoder, into our dataframe.

In [24]:
le = LabelEncoder()
df['result'] = le.fit_transform(df['diagnosis'])
df

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         M        17.99         10.38          122.80     1001.0   
1      842517         M        20.57         17.77          132.90     1326.0   
2    84300903         M        19.69         21.25          130.00     1203.0   
3    84348301         M        11.42         20.38           77.58      386.1   
4    84358402         M        20.29         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
564    926424         M        21.56         22.39          142.00     1479.0   
565    926682         M        20.13         28.25          131.20     1261.0   
566    926954         M        16.60         28.08          108.30      858.1   
567    927241         M        20.60         29.33          140.10     1265.0   
568     92751         B         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     ...  area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0    ...      2019.0           0.16220            0.66560           0.7119   
1    ...      1956.0           0.12380            0.18660           0.2416   
2    ...      1709.0           0.14440            0.42450           0.4504   
3    ...       567.7           0.20980            0.86630           0.6869   
4    ...      1575.0           0.13740            0.20500           0.4000   
..   ...         ...               ...                ...              ...   
564  ...      2027.0           0.14100            0.21130           0.4107   
565  ...      1731.0           0.11660            0.19220           0.3215   
566  ...      1124.0           0.11390            0.30940           0.3403   
567  ...      1821.0           0.16500            0.86810           0.9387   
568  ...       268.6           0.08996            0.06444           0.0000   

     concave points_worst  symmetry_worst  fractal_dimension_worst  \
0                  0.2654          0.4601                  0.11890   
1                  0.1860          0.2750                  0.08902   
2                  0.2430          0.3613                  0.08758   
3                  0.2575          0.6638                  0.17300   
4                  0.1625          0.2364                  0.07678   
..                    ...             ...                      ...   
564                0.2216          0.2060                  0.07115   
565                0.1628          0.2572                  0.06637   
566                0.1418          0.2218                  0.07820   
567                0.2650          0.4087                  0.12400   
568                0.0000          0.2871                  0.07039   

     Unnamed: 32  target  result  
0            NaN       1       1  
1            NaN       1       1  
2            NaN       1       1  
3            NaN       1       1  
4            NaN       1       1  
..           ...     ...     ...  
564          NaN       1       1  
565          NaN       1       

In [25]:
df.isnull().sum()

id                           0
diagnosis                    0
radius_mean                  0
texture_mean                 0
perimeter_mean               0
area_mean                    0
smoothness_mean              0
compactness_mean             0
concavity_mean               0
concave points_mean          0
symmetry_mean                0
fractal_dimension_mean       0
radius_se                    0
texture_se                   0
perimeter_se                 0
area_se                      0
smoothness_se                0
compactness_se               0
concavity_se                 0
concave points_se            0
symmetry_se                  0
fractal_dimension_se         0
radius_worst                 0
texture_worst                0
perimeter_worst              0
area_worst                   0
smoothness_worst             0
compactness_worst            0
concavity_worst              0
concave points_worst         0
symmetry_worst               0
fractal_dimension_worst      0
Unnamed:

### By checking the Null values count, we need to drop 'Unnamed: 32' also along with 'id', as both of them are not required for our model.

In [ ]:
df.drop(['Unnamed: 32'], axis=1, inplace=True)

In [35]:
df.drop(['diagnosis'], axis=1, inplace=True)

In [41]:
df.drop(['id'], axis=1, inplace=True)

In [42]:
df.shape

(569, 31)

Finally, after applying data cleaning & preprocessing, we got a dataset with rows=569 & columns=31


In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   radius_mean              569 non-null    float64
 1   texture_mean             569 non-null    float64
 2   perimeter_mean           569 non-null    float64
 3   area_mean                569 non-null    float64
 4   smoothness_mean          569 non-null    float64
 5   compactness_mean         569 non-null    float64
 6   concavity_mean           569 non-null    float64
 7   concave points_mean      569 non-null    float64
 8   symmetry_mean            569 non-null    float64
 9   fractal_dimension_mean   569 non-null    float64
 10  radius_se                569 non-null    float64
 11  texture_se               569 non-null    float64
 12  perimeter_se             569 non-null    float64
 13  area_se                  569 non-null    float64
 14  smoothness_se            5

In [44]:
df.describe()

radius_mean  texture_mean  perimeter_mean    area_mean  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       symmetry_mean  fractal_dimension_mean  ...  texture_worst  \
count     569.000000              569.000000  ...     569.000000   
mean        0.181162                0.062798  ...      25.677223   
std         0.027414                0.007060  ...       6.146258   
min         0.106000                0.049960  ...      12.020000   
25%         0.161900                0.057700  ...      21.080000   
50%         0.179200                0.061540  ...      25.410000   
75%         0.195700                0.066120  ...      29.720000   
max         0.304000                0.097440  ...      49.540000   

       perimeter_worst   area_worst  smoothness_worst  compactness_worst  \
count       569.000000   569.000000        569.000000         569.000000   
mean        107.261213   880.583128          0.132369           0.254265   
std          33.602542   569.356993          0.022832           0.157336   
min          50.410000   185.200000          0.071170           0.027290   
25%          84.110000   515.300000          0.116600           0.147200   
50%          97.660000   686.500000          0.131300           0.211900   
75%         125.400000  1084.000000          0.146000           0.339100   
max         251.200000  4254.000000          0.222600           1.058000   

       concavity_worst  concave points_worst  symmetry_worst  \
count       569.000000            569.000000      569.000000   
mean          0.272188              0.114606        0.290076   
std           0.208624              0.065732        0.061867   
min           0.000000              0.000000        0.156500   
25%           0.114500              0.064930        0.250400   
50%           0.226700              0.099930        0.282200   
75%           0.382900              0.161400        0.317900   
max           1.252000              0.291000        0.663800   

       fractal_dimension_worst      result  
count               569.000000  569.000000  
mean                  0.083946    0.372583  
std                   0.018061    0.483918  
min                   0.055040    0.000000  
25%                   0.071460    0.000000  
50%                   0.080040    0.000000  
75%                   0.092080    1.000000  
max                   0.207500    1.000000  

[8 rows x 31 columns]

By this describe() method, we get to know useful statistics about our features separately.

In [45]:
df['result'].value_counts()  #357=Benign=0, 212=Malignant=1

0    357
1    212
Name: result, dtype: int64

## Class 0 = Benign = total 357 ___&___ Class 1 = Malignant = total 212

In [46]:
df.groupby('result').mean()

radius_mean  texture_mean  perimeter_mean   area_mean  \
result                                                          
0         12.146524     17.914762       78.075406  462.790196   
1         17.462830     21.604906      115.365377  978.376415   

        smoothness_mean  compactness_mean  concavity_mean  \
result                                                      
0              0.092478          0.080085        0.046058   
1              0.102898          0.145188        0.160775   

        concave points_mean  symmetry_mean  fractal_dimension_mean  ...  \
result                                                              ...   
0                  0.025717       0.174186                0.062867  ...   
1                  0.087990       0.192909                0.062680  ...   

        radius_worst  texture_worst  perimeter_worst   area_worst  \
result                                                              
0          13.379801      23.515070        87.005938   558.899440   
1          21.134811      29.318208       141.370330  1422.286321   

        smoothness_worst  compactness_worst  concavity_worst  \
result                                                         
0               0.124959           0.182673         0.166238   
1               0.144845           0.374824         0.450606   

        concave points_worst  symmetry_worst  fractal_dimension_worst  
result                                                                 
0                   0.074444        0.270246                 0.079442  
1                   0.182237        0.323468                 0.091530  

[2 rows x 30 columns]

### After finding the number or counts of both the classes, we obtained average results for the features. Now we need to divide our data into input_to_model (x) & target (y).

In [48]:
x = df.drop(columns = 'result', axis=1)
y = df['result']

In [ ]:
print(x)

In [50]:
print(y)

0      1
1      1
2      1
3      1
4      1
      ..
564    1
565    1
566    1
567    1
568    0
Name: result, Length: 569, dtype: int32


### To create a model we need to divide the data set into 4-main categories as shown below:

In [51]:
# We'll use 20% data for testing

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=2)


In [53]:
print(x.shape, x_train.shape, x_test.shape, y_train.shape, y_test.shape)

(569, 30) (455, 30) (114, 30) (455,) (114,)


# Model Training : Support Vector Machine

We will create our Support Vector Machine model using scikit-learn.
SVM is used as classification techniques which first attempts to find Hyperplane, which can separate objects of different classes and then leaves maximum marigin from the nearest points of each class to the separating hyperplane.

In [56]:
model = svm.SVC(kernel = 'linear', gamma = 'auto', C=0.3)

Here kernel, gamma & c are explained below:
1. kernel method is a class of algorithms for pattern analysis.
2. gamma parameter determines the distance a single data sample exerts influence or it is said to adjust the curvature of the decision boundary.
3. C determines penalty or how many samples are allowed to be placed in different classes.

In [57]:
model.fit(x_train, y_train)

SVC(C=0.3, gamma='auto', kernel='linear')

In [58]:
y_predicted = model.predict(x_test)

By doing this we can find the difference between the actual data (y_test) & predicted data (y_predicted.

Further, we will evaluate the results & check the precision and f1 score.

## Results, Precision and F1 score

In [60]:
print(classification_report(y_test, y_predicted))

              precision    recall  f1-score   support

           0       0.94      0.96      0.95        69
           1       0.93      0.91      0.92        45

    accuracy                           0.94       114
   macro avg       0.94      0.93      0.94       114
weighted avg       0.94      0.94      0.94       114



Here,
1. Precision: Precision measures the percentage of true positives (correctly predicted positive instances) out of all predicted positives. It gives an indication of how accurate the model's positive predictions are.

    Precision = True Positives / (True Positives + False Positives)


2. Recall: Recall measures the percentage of true positives out of all actual positives. It gives an indication of how well the model is able to identify positive instances.

    Recall = True Positives / (True Positives + False Negatives)


3. F1-score: The F1-score is the harmonic mean of precision and recall, and is a balanced metric that takes both into account. It provides a single measure of a model's performance.

    F1-score = 2 * (Precision * Recall) / (Precision + Recall)


4. support means instances.

### Class 0 = Benign 
### Class 1 = Malignant

# Here 114 is number of Test-Data.
# Prediction by model is 69 (Benign) + 45(Malignant).
# This is the result of our dataset with accuracy of 94%.


# -
# -
# -
# -
### Thank You
### Prepared By: Mayur Kumar Sharma / DW-32 Batch / GUVI